
\begin{acronym}[MPC]

\acro{ABR}{Antibiotic Resistant}
\acro{agMLST}{accessory-genome MLST}
\acro{AI}{Artificial Intelligence}
\acro{ALM}{Allele Larger than Mode}
\acro{AMR}{Antimicrobial Resistance}
\acro{ANI}{Average Nucleotide Identity}
\acro{AP-PCR or RAPD}{Arbitrarily Primed PCR}
\acro{API}{Application Programming Interface}
\acro{AR}{Adjusted Rand}
\acro{ASM}{Allele Smaller than Mode}
\acro{AW}{Adjusted Wallace}
\acro{BIGSdb}{Bacterial Isolate Genome Sequence Database}
\acro{BLAST}{Basic Local Alignment Search Tool}
\acro{BLASTn}{Nucleotide BLAST}
\acro{BLASTp}{Protein BLAST}
\acro{BLASTx}{translated nucleotide BLAST}
\acro{bp}{base pairs}
\acro{BPPL}{Bacterial Priority Pathogens List}
\acro{BSR}{BLAST Score Ratio}
\acro{CC}{Clonal Complex}
\acro{CDC}{Centers for Disease Control and Prevention}
\acro{cDNA}{complementary DNA}
\acro{CDS}{coding DNA sequence}
\acro{cgMLST}{core-genome MLST}
\acro{Chewie-NS}{Chewie Nomenclature Server}
\acro{CLI}{Command Line Interface}
\acro{CPU}{Central Processing Unit}
\acro{ddNTPs}{dideoxynucleotides}
\acro{DNA}{Deoxyribonucleic acid}
\acro{dNTPs}{deoxynucleotides}
\acro{ECDC}{European Centre for Disease Prevention and Control}
\acro{EFSA}{European Food Safety Authority}
\acro{ENA}{European Nucleotide Archive}
\acro{EU/EEA}{European Union and the European Economic Area}
\acro{EXC}{Exact Match}
\acro{FAIR}{Findable Accessible Interoperable Reusable}
\acro{FCT}{Fundação para a Ciência e Tecnologia}
\acro{FWD}{Food and Waterborne disease}
\acro{GAS}{Lanceﬁeld group A \textit{Streptococcus}}
\acro{GbG}{Gene-by-Gene}
\acro{GIGO}{Garbage In, Garbage Out}
\acro{GPU}{Graphics Processing Unit}
\acro{GUI}{General User Interface}
\acro{HGT}{Horizontal Gene Transfer}
\acro{HR}{Homologous Recombination}
\acro{HTML}{HyperText Markup Language}
\acro{HTS}{high-throughput sequencing}
\acro{HTTPS}{Hypertext Transfer Protocol Secure}
\acro{kb}{kilobases}
\acro{LOTSC}{LOTSC}
\acro{MALDI-TOF}{matrix-assisted laser desorption/ionization-time-of-flight}
\acro{Mb}{megabases}
\acro{MB}{megabyte}
\acro{MGEs}{Mobile Genetic Elements}
\acro{MIC}{minimum inhibitory concentration}
\acro{MLEE}{multilocus enzyme electrophoresis}
\acro{MLST}{Multilocus Sequence Typing}
\acro{MLVA}{multiple-locus variable number tandem repeat analysis}
\acro{MPS}{massive parallel sequencing}
\acro{MSA}{Multiple Sequence Alignment}
\acro{MST}{Minimum-spanning-tree}
\acro{MTC}{Mycobacterium tuberculosis complex}
\acro{NCBI}{National Center for Biotechnology Information}
\acro{NIPH}{Non-Informative Paralogous Hit}
\acro{NIPHEM}{Non-Informative Paralogous Hit Exact Match}
\acro{NJ}{Neighbor-Joining}
\acro{NS}{Nomenclature Server}
\acro{ONT}{Oxford Nanopore Technologies}
\acro{ORF}{open-reading frame}
\acro{OS}{Operating System}
\acro{PacBio}{Pacific Biosciences}
\acro{PAMA}{Paralogous Match}
\acro{PCR}{Polymerase Chain Reaction}
\acro{PE}{paired-end}
\acro{PFGE}{Pulse Field Gel Electrophoresis}
\acro{pgMLST}{pan-genome MLST}
\acro{PLOT}{Possible Locus On the Tip}
\acro{PLOT3}{PLOT 3'-end}
\acro{PLOT5}{PLOT 5'-end}
\acro{PopPUNK}{POPulation Partitioning Using Nucleotide Kmers}
\acro{PVL}{Panton-Valentine leukocidin}
\acro{qPCR}{Real-Time Quantitative PCR}
\acro{RAM}{Random Access Memory}
\acro{RD}{Research and Development}
\acro{RDF}{Resource Description Framework}
\acro{RE}{restriction enzyme}
\acro{REP-PCR}{repetitive sequencing-based PCR}
\acro{REST}{Representational State Transfer}
\acro{RF}{restriction fragment}
\acro{RFLP}{Restriction Fragment Length Polymorphism}
\acro{rMLST}{Ribosomal MLST}
\acro{rps}{ribosomal protein subunit}
\acro{rRNA}{ribossomal ribonucleic acid}
\acro{rST}{ribosomal ST}
\acro{SB}{Synteny Block}
\acro{SBS}{sequencing by synthesis}
\acro{SID}{Simpson’s index of diversity}
\acro{SMRT}{Single Molecule Real-Time Sequencing}
\acro{SNA}{Synteny Network Analysis}
\acro{SNP}{Single Nucleotide Polymorphism}
\acro{SNV}{Single Nucleotide Variant}
\acro{SPARQL}{SPARQL Protocol and RDF Query Language}
\acro{ssDNA}{single-stranded DNA}
\acro{ST}{Sequence Type}
\acro{SVG}{Scalable Vector Graphics}
\acro{TSV}{Tab-Separated Values}
\acro{UI}{User Interface}
\acro{UPGMA}{unweighted pair group method with arithmetic means}
\acro{VNTR}{variable number tandem repeats}
\acro{wg/cgMLST}{whole- and core-genome MLST}
\acro{wgMLST}{whole-genome MLST}
\acro{WGS}{Whole Genome Sequencing}
\acro{WHO}{World Health Organization}
\acro{WS}{Web Service}
\acro{WSGI}{Web Server Gateway Interface}

\end{acronym}
