


The emergence of accurate and cost-efficient high-throughput DNA sequencing technologies has revolutionized biological sciences. These technologies have been widely adopted by research and public health institutions, providing increased resolution in a wide range of applications to complement or replace more classical phenotypic and molecular assays. The wealth and complexity of the generated data demanded greater storage capacity and improved computational methods for data analysis to make sense of the data. These demands potentiated a tremendous growth in the field of bioinformatics, which has become an integral part in omics approaches such as genomics. In bacterial genomics, in particular, bioinformatics methods have become essential to characterize bacterial strains, providing higher resolution in areas such as infectious disease surveillance and outbreak investigation, as well as for the study of the structure and evolution of bacterial populations. Whole genome sequencing allows for a purely sequence-based approach for bacterial characterization, enabling targeted approaches to identify genes of interest or the analysis of the full gene diversity. The increased availability of complete or nearly complete bacterial genome assemblies allowed researchers to study the structure and variability of bacterial genomes and encouraged the development of approaches for high-resolution bacterial typing, such as gene-by-gene and SNP-based methods. Although both approaches have been extensively applied, either separately or in combination, in comparative genomics to study the diversity of bacterial populations, gene-by-gene methods such as whole- and core-genome multilocus sequence typing have been adopted more frequently by research and public health institutions for the surveillance and outbreak investigation of bacterial pathogens.




The present thesis presents met

