
\begin{acronym}[MPC]

\acro{AMR}{Antimicrobial Resistance}
\acro{ANI}{Average Nucleotide Identity}
\acro{API}{Application Programming Interface}
\acro{ALM}{Allele Larger than Mode}
\acro{AR}{adjusted Rand}
\acro{ASM}{Allele Smaller than Mode}
\acro{AW}{adjusted Wallace}
\acro{bp}{basepairs}
\acro{BLAST}{Basic Local Alignment Search Tool}
\acro{BLASTn}{Nucleotide BLAST}
\acro{BLASTp}{Protein BLAST}
\acro{BLASTx}{translated nucleotide BLAST}
\acro{BSR}{BLAST Score Ratio}
\acro{agMLST}{accessory-genome MLST}
\acro{cgMLST}{core-genome MLST}
\acro{pgMLST}{pan-genome MLST}
\acro{wgMLST}{whole-genome MLST}
\acro{wg/cgMLST}{whole- and core-genome Multilocus Sequence Typing}
\acro{CDC}{Centers for Disease Control and Prevention}
\acro{CDS}{coding DNA sequence}
\acro{CDSs}{coding DNA sequences}
\acro{Chewie-NS}{Chewie Nomenclature Server}
\acro{ddNTP}{dideoxynucleotide triphosphate}
\acro{DNA}{Deoxyribonucleic acid}
\acro{dNTP}{deoxynucleotide triphosphate}
\acro{ECDC}{European Centre for Disease Prevention and Control}
\acro{EFSA}{European Food Safety Authority}
\acro{ENA}{European Nucleotide Archive}
\acro{EXC}{Exact Match}
\acro{FAIR}{Findable Accessible Interoperable Reusable}
\acro{FCT}{Fundação para a Ciência e Tecnologia}
\acro{FWD}{Food and Waterborne disease}
\acro{GIGO}{Garbage In, Garbage Out}
\acro{GAS}{Lanceﬁeld group A \textit{Streptococcus}}
\acro{GbG}{Gene-by-Gene}
\acro{GPP}{Global Priority Pathogens}
\acro{HTML}{HyperText Markup Language}
\acro{HTTPS}{Hypertext Transfer Protocol Secure}
\acro{HTS}{high-throughput sequencing}
\acro{LOTSC}{LOTSC}
\acro{MB}{megabyte}
\acro{MGEs}{Mobile Genetic Elements}
\acro{MLST}{Multilocus Sequence Typing}
\acro{MSA}{Multiple Sequence Alignment}
\acro{MST}{Minimum-spanning-tree}
\acro{NCBI}{National Center for Biotechnology Information}
\acro{NJ}{Neighbor-Joining}
\acro{NIPH}{Non-Informative Paralogous Hit}
\acro{NIPHEM}{Non-Informative Paralogous Hit Exact Match}
\acro{NS}{Nomenclature Server}
\acro{OLC}{Overlap-Layout Consensus}
\acro{ONT}{Oxford Nanopore Technologies}
\acro{ORF}{open-reading frame}
\acro{ORFs}{open-reading frames}
\acro{PA}{presence-absence}
\acro{PacBio}{Pacific Biosciences}
\acro{PAMA}{Paralogous Match}
\acro{PCR}{Polymerase Chain Reaction}
\acro{PLOT}{Possible Locus On the Tip}
\acro{PLOT3}{PLOT 3'-end}
\acro{PLOT5}{PLOT 5'-end}
\acro{PFGE}{Pulse Field Gel Electrophoresis}
\acro{PopPUNK}{POPulation Partitioning Using Nucleotide Kmers}
\acro{qPCR}{Real-Time Quantitative PCR}
\acro{RAM}{Random Access Memory}
\acro{REST}{Representational State Transfer}
\acro{RDF}{Resource Description Framework}
\acro{RNA}{Ribonucleic acid}
\acro{rRNA}{ribossomal RNA}
\acro{RT-PCR}{Reverse Transcription Polymerase Chain Reaction}
\acro{SID}{Simpson’s index of diversity}
\acro{SMRT}{Single Molecule Real-Time Sequencing}
\acro{SNP}{Single Nucleotide Polymorphism}
\acro{SNPs}{Single Nucleotide Polymorphisms}
\acro{SNV}{Single Nucleotide Variant}
\acro{SNVs}{Single Nucleotide Variants}
\acro{SPARQL}{SPARQL Protocol and RDF Query Language}
\acro{ST}{Sequence Type}
\acro{SVG}{Scalable Vector Graphics}
\acro{TSV}{Tab-Separated Values}
\acro{wg/cgMLST}{whole- and core-genome MLST}
\acro{WGS}{Whole Genome Sequencing}
\acro{WHO}{World Health Organization}
\acro{WS}{Web Service}
\acro{WSGI}{Web Server Gateway Interface}
\acro{UI}{User Interface}
\acro{UPGMA}{unweighted pair group method with arithmetic means}

\end{acronym}
