This thesis describes the development and application of improved bioinformatics methods for bacterial typing based on whole genome sequencing, more specifically gene-by-gene approaches such as \ac{wg/cgMLST}. wg/cgMLST is widely used by research and public health institutions for the surveillance and study of the population diversity of bacterial pathogens. Therefore, improving wg/cgMLST methodologies is of relevance, since any improvements can be more easily adopted by the scientific community and public health authorities, having a more direct impact on research and public health.

The thesis is organized into five chapters.

\textbf{Chapter 1} is a general introduction to bacterial typing methods, starting with a brief explanation of some of the classical biochemical and molecular methods used in the laboratory and building on that to transition to the WGS-based approaches possible by the DNA sequencing revolution. The main WGS-based approaches for bacterial typing are explained, although gene-by-gene approaches, such as wg/cgMLST, are explained more thoroughly.

The general introduction is followed by the body of the thesis, composed of three selected manuscripts that present methods and results that aim to improve the efficiency, accuracy, and interoperability of wg/cgMLST. Each manuscript has its own dedicated chapter. There is a high degree of intersection between the subjects of each chapter, and the work of each chapter contributes in part or is instrumental to advance the work presented in the other chapters.

\textbf{Chapter 2} describes the implementation and evaluation of a bioinformatics tool for wg/cgMLST, chewBBACA 3. chewBBACA 3 is a reimplementation of chewBBACA's first published version and was designed to be a complete solution for wg/cgMLST, offering improved and new functionalities that considerably expand chewBBACA's capabilities. It enables fast and efficient schema creation from multiple sources and allele calling with modest computational resources to meet current and future data processing demands. The speed and scalability of the allele calling opens the possibility of large-scale wg/cgMLST for more users that want to study the population diversity of a species of interest based on large datasets or in surveillance and outbreak scenarios where a timely analysis is crucial. To take advantage of the extensive schema and results data produced, chewBBACA 3 includes modules for a comprehensive analysis of the schemas and allele calling results. These modules produce interactive and easily shareable reports based on an upgraded version of the user interface developed for the Web service developed in \textbf{Chapter 3}, allowing users of any level of expertise to explore the results and make an informed decision more easily without having to resort to multiple solutions or create custom solutions for the analysis of the results.

\textbf{Chapter 3} presents Chewie-NS, a Nomenclature Server that stores wg/cgMLST schemas and allows for local and private wg/cgMLST analysis based on a common allelic nomenclature. The clear and simple interface of Chewie-NS allows users to easily find, explore loci diversity, and download schemas for species of interest. For users who are more tech-savvy, the API provides access to all schema data, either through Swagger UI or programmatically. The schemas downloaded from Chewie-NS can be used to perform local an private allele calling, minimizing scalability issues if data analysis was centralized in the web server and allowing users operating under stricter data privacy policies to still take advantage of the service. This contrasts with other wg/cgMLST platforms, which typically centralize data analysis, requiring users to submit their data and establishing queues to process data submissions based on available computational resources. The local schemas downloaded from Chewie-NS can be synchronized with the remote versions to retrieve novel alleles added to the remote schemas and contribute novel alleles identified locally if desired. The synchronization process allows one to update both schema versions and the allelic nomenclature to enable results comparison even if data analysis is not centralized. To provide functionalities for users to use the schemas deposited in Chewie-NS, a set of modules for chewBBACA, presented in \textbf{Chapter 2}, was implemented to allow the upload, download, and synchronization of the schemas. This integration provides access to the schemas deposited in Chewie-NS, allowing users to perform local allele calling and compare their results based on a common allelic nomenclature, while also allowing users to take advantage of chewBBACA's powerful analytic capabilities.

\textbf{Chapter 4} describes the creation and evaluation of an annotated wgMLST schema for \textit{S. pyogenes}, a major human pathogen with high genetic diversity. The target loci for the wgMLST schema were defined based on a dataset of high-quality complete genomes and the schema was populated through allele calling with multiple datasets representing the known diversity of \textit{S. pyogenes}. This schema was refined based on automatic annotations and the suggestions of a domain expert. The refined wgMLST schema, comprising 3,044 loci, allows for high-resolution typing of \textit{S. pyogenes} for the study of diverse datasets and in an outbreak context, displaying performance comparable to SNP-based methods. The development of the wgMLST schema benefited from the novel functionalities that were being implemented in chewBBACA concomitantly. The schema refining process also revealed key limitations and challenges of scaling current gene-by-gene approaches, often based only on cgMLST, to wgMLST, and provided crucial information to guide the development of some functionalities included in chewBBACA 3, presented in \textbf{Chapter 2}. The wgMLST schema was deposited in Chewie-NS, presented in \textbf{Chapter 3}.

\textbf{Chapter 5} corresponds to the general discussion. In this chapter, the results presented in each chapter of the body of the thesis are summarized and discussed in terms of their advantages, disadvantages, and the potential to be further improved.
